src/TextInputWidget.cc

```diff
@@ -135,28 +135,42 @@ FilteredTextEdit::canInsertFromMimeData(const QMimeData *source) const
 void
 FilteredTextEdit::insertFromMimeData(const QMimeData *source)
 {
-        if (source->hasImage()) {
-                const auto formats = source->formats();
-                const auto idx     = formats.indexOf(
-                  QRegularExpression{"image/.+", QRegularExpression::CaseInsensitiveOption});
-
-                // Note: in the future we may want to look into what the best choice is from the
-                // formats list. For now we will default to PNG format.
-                QString type = "png";
-                if (idx != -1) {
-                        type = formats.at(idx).split('/')[1];
+        const auto formats = source->formats().filter("/");
+        const auto image   = formats.filter("image/", Qt::CaseInsensitive);
+        const auto audio   = formats.filter("audio/", Qt::CaseInsensitive);
+        const auto video   = formats.filter("video/", Qt::CaseInsensitive);
+
+        if (!image.empty()) {
+                showPreview(source, image);
+        } else if (!audio.empty()) {
+                showPreview(source, audio);
+        } else if (!video.empty()) {
+                showPreview(source, video);
+        } else if (source->hasUrls()) {
+                QString path;
+                if (QFileInfo{source->text()}.exists()) {
```

Comment:
Why is this branch needed? Shouldn't processing the URLs directly work in all cases where this would?

Reply:
It's there to short-circuit the rest of the code in case `source` contains URLs and also a plain version of the file path (without the `file://` prefix).

Reply:
Why is that useful? Processing the URLs is not hard, and is far more likely to be reliable; there's no guarantee that the text is a path.

Reply:
It's a premature optimization. On second thought, it is not really that useful. Will change.